### ISSUE TITLE
Feature Request: Please support multithreading analysis!

### ISSUE THREAD

**Issue by ytrezq** [ConsenSys/mythril#1260]
Myth can run during hours but always using thread.

Now that the official python implementation has a [workaround for the gil](https://www.python.org/dev/peps/pep-0554/), is there a way for Myhril to benefit from it?

**Comment by norhh** [comment]
That sounds great. We will try to use this.

**Comment by ytrezq** [comment]
That also means that right now, all new code change should ensure thread safety.

**Comment by ytrezq** [comment]
An alternative would be to simply use the Numba packages which can run threads without using the gil at all.
It would also dramatically speed up single threaded mode.

**Comment by ytrezq** [comment]
A large portion of the time spent when analysing contract is actually performed in C code in the Z3 solver.
For example, when analysing `0xbb9bc244d798123fde783fcc1c72d3bb8c189413` using `-t 1` limit from mainnet, for 2779 queries completing in 2200 seconds on my machine, it took 129 seconds in `Z3_optimize_check` with only 981 calls and 1798 calls to `Z3_solver_check_assumptions` taking 449 seconds (and apparently without calling anything else in both cases).

@norhh so the question: is there no way to do something else not depending on the result while those functions are called? Or since I’m seeing often a thread being spawn it’s already happening in multithread mode?

**Comment by norhh** [comment]
Yes, it was possible. But we implemented multithreading previously but there was a slowdown due to GIL. We can't exactly use multiprocessing as Z3 isn't serialisable friendly. Switching to other versions might help with multi threading.

**Comment by ytrezq** [comment]
@norhh I was talking regardless of the Gil of course. Please note Numba doesn’t rely on serialization and can run even old Python versions.

So what can be done while z3 works on `Z3_optimize_check` or `Z3_solver_check_assumptions`? Something that [doesn’t access the Python api if possible](https://numba.pydata.org/numba-doc/dev/glossary.html#term-nopython-mode).

**Comment by JoranHonig** [comment]
@ytrezq The previous approach to multithreading was based on a previous design of laser and mythril, with the current architecture multithreaded symbolic execution would look a little bit different.

Right now the process can generally be seen like this:
- There is a single worker thread ( the main thread )
- There is a worklist of states that need to be processed
- The worker fetches a state from the worklist according to some strategy
- The worker computes the successor state(s) for that state and adds them to the worklist

We can parallelize this by having multiple worker threads. (there will need to be some concurrency control added for some of the features & plugins in laser)

I'm not sure what the limitations of Numba are, so I'm not sure how that affects our ability to paralellize things.

**Comment by ytrezq** [comment]
@JoranHonig: in the case of Numba here’s the 2 limitations:
- The function and it’s subfunctions (including in dependent modules) must not rely on the Python C api (not that it doesn’t code C code if the C/C++ code being called doesn’t).
- And the native types of all values in the function and subfunctions (including in dependent modules) can be inferred.

That’s it! and then the process to parallelize is just to add this function decorator:
`@njit(parallel=True,nopython=True)` before function declaration and it will automatically vectorize it and parallelize it (taking care of thread safety). And as the name jit and nopython=True implies, even if the result can’t be vectorized or panellized, it will run far faster than in normal mode. Numba is even compatible with Python3.4 which mean the approach would work right now.

 And basically, the [implemented workaround](https://www.python.org/dev/peps/pep-0554/) which can be tested right now in the latest version works exactly like multiprocess mode except it remove the ipc overhead.

But what I’m seeing is between 20% to 55% of the time spent in the main thread is performed in the solver. That time spent doesn’t requires the python interpreter at all which means it doesn’t even care about what the gil is about.
We already have part which is performed in a separate thread (I failed to find which one but it represent a 10% increase over single thread mode) from time to time even if the program is bound to main thread speed, so why not decrease a one day analysis time by 7 to 10 hours by spawning works which isn’t related to the solver off the main thread (leaving solver work in the main thread).

**Comment by ytrezq** [comment]
@norhh yes I known the codebase is no longer related. But neverless I’m interested about getting more details about in which functions and files it was done.

**Comment by ytrezq** [comment]
@norhh any chance this get into mythx?

**Comment by norhh** [comment]
@ytrezq We are currently working on other issues, we will inform after it's decided.

**Comment by ytrezq** [comment]
@norhh might be a good solution for your servers which would allow the previous attempt to just work. I found a gil free version of Python3.6 https://github.com/larryhastings/gilectomy
